src/Combinatorics/Matroids/matroids.jl

```diff
@@ -0,0 +1,825 @@
+export
+    Matroid, matroid_groundset, 
+    matroid_from_revlex_basis_encoding,
+    matroid_from_bases, matroid_from_nonbases, matroid_from_circuits, matroid_from_hyperplanes,
+    matroid_from_matrix_columns, matroid_from_matrix_rows,
+    cycle_matroid, bond_matroid, cocycle_matroid,
+    dual_matroid, direct_sum, restriction, deletion, contraction, minor,
+    principal_extension, free_extension, series_extension, parallel_extension,
+    uniform_matroid, fano_matroid, non_fano_matroid, pappus_matroid, non_pappus_matroid, vamos_matroid,
+    all_subsets_matroid, projective_plane, projective_geometry, affine_geometry
+
+
+################################################################################
+##  Constructing
+################################################################################
+ElementType = Union{IntegerUnion,Char,String}
+GroundsetType = Union{AbstractVector, AbstractSet} 
+
+
+struct Matroid
+    pm_matroid::Polymake.BigObject
+    groundset::GroundsetType # groundset of the matroid 
+    gs2num::Dict{Any, IntegerUnion}# dictionary to map the groundset to the integers from 1 to its size
+end
+
+pm_object(M::Matroid) = M.pm_matroid
+
+function Base.show(io::IO, M::Matroid)
+    r = rank(M)
+    n = length(M.groundset)
+    print(io, "Matroid of rank $(r) on $(n) elements")
+end
+
+# function that generates the dictionary which maps the groundset to integers
+function create_gs2num(E::GroundsetType)
+    gs2num = Dict{Any,IntegerUnion}()
+    [gs2num[E[i]] = i for i in 1:length(E)]
+    return gs2num
+end
```

Comment:
This is abusing the side effect of a list comprehension, at the cost of creating and discarding a `Vector` equal to `1:length(E)`. I'd suggest to instead write this, which is more efficient and IMO also more readable (but of course that's subjective)
```suggestion
create_gs2num(E::GroundsetType) = Dict{Any,IntegerUnion}(E[i] => i for i in 1:length(E))
```